flow/include/flow/OwningResource.h

```diff
@@ -0,0 +1,119 @@
+/*
+ * OwningResource.h
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2013-2022 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FLOW_OWNING_REOSURCE_H
+#define FLOW_OWNING_REOSURCE_H
+
+#include "flow/FastRef.h"
+
+// Consider the following situation:
+//
+//   1. An ACTOR A0 allocates an object O
+//   2. A0 spawns another ACTOR A1, which depends on O
+//   3. A0 triggers A1 and then terminates, destroying O
+//   4. Since A1 is triggered by A0 while not knowing A0 is terminated and O is released, it would cause a SEGV error.
+//
+// In this header file, two classes
+//
+//   * ResourceOwningRef
+//   * ResourceWeakRef
+//
+// are provided. The ResourceOwningRef is the reference that "holds" the resource, When it is destructed, the resource
+// is also released; while ResourceWeakRef is the reference that "weakly holds" the resource. Before each access, it is
+// the user's responsibility to verify if the resource is still available, via the available() method.
+//
+// With the two classes, the issue above can be solved by:
+//
+//    1. A0 allocates the object O via ResourceOwningRef
+//    2. A0 forwards O to A1, via ResourceWeakRef
+//    3. Every time A1 accesses O, it will verify if the resource is still available.
+//    4. When A0 terminates, O is released and all ResourceWeakRef available() call will report the resource is not
+//       available anymore, preventing the SEGV error being raised.
+
+namespace details {
+
+// The class holding the pointer to the resource.
+// SOMEDAY: Think using std::unique_ptr
+template <typename T>
+struct Resource : public ReferenceCounted<Resource<T>>, NonCopyable {
+	T* resource;
+
+	Resource(T* resource_) : resource(resource_) {}
+	~Resource() { delete resource; }
+
+	void reset(T* resource_) {
+		delete resource;
+		resource = resource_;
+	}
+};
+
+template <typename T>
+class ResourceRef {
+protected:
+	Reference<Resource<T>> resourceRef;
+
+public:
+	ResourceRef(const Reference<Resource<T>>& ref) : resourceRef(ref) {}
+	ResourceRef(Reference<Resource<T>>&& ref) : resourceRef(std::move(ref)) {}
+	ResourceRef& operator=(const Reference<Resource<T>>& ref) {
+		resourceRef = ref.resourceRef;
+		return *this;
+	}
+	ResourceRef& operator=(Reference<Resource<T>>&& ref) {
+		resourceRef = std::move(ref);
+		return *this;
+	}
+
+	T* operator->() { return resourceRef->resource; }
```

Comment:
To match the behaviour of `operator*`, this should throw an error if the resource is invalid.

Reply:
For `dynamic_cast` they are treated differently, but `operator->` should never be used on a `nullptr`.

Reply:
I think this should be enforced by always checking the `available` when accessing a pointer-liked object, as it is the caller's responsibility to verify if a pointer is `nullptr` or not before using it. But regarding to @ajbeamon's suggestion, I would think adding another layer which throws `resource_not_available` exception when the weak reference is used against released resources.

Reply:
I'm not sure if introducing a new error would address my concern specifically. My main worry is that in combination with an actor that has wait statements in it, it is probably too easy to misuse this. If the typical use-case is to handle a race condition between an actor being canceled and the argument to it being destroyed, then we could potentially assume that the argument being destroyed is a signal to cancel the actor and do so.

Reply:
In that scenario, though, this could be reframed as a special purpose concept for actors (e.g. `WeakActorSelfPointer`, or something like that).

Reply:
The reason I do not feel throwing an `operation_cancelled`, or `actor_cancelled`, is that doing so seems like escalating error on one variable to the whole ACTOR. Also, regarding to `WeakActorSelfPointer`, I am not sure if it, on the opposite, restricting the use of the class to the extent of "self" pointer.

Reply:
It doesn't necessarily have to be "self", but I think this use-case could warrant a special purpose construct since it is a common pattern in our codebase. And if we had that construct, it might be more applicable here than the generic version you've implemented.

Reply:
I think in this case it is simple -- just another layer of subclassing should fulfill this requirement.